sw/device/lib/dif/dif_kmac.c

```diff
@@ -311,13 +311,22 @@ dif_kmac_result_t dif_kmac_absorb(dif_kmac_t *kmac, const void *msg, size_t len,
     return kDifKmacError;
   }
 
-  // Copy the message one byte at a time.
-  // This could be sped up copying a word at a time but be careful
-  // about message endianness (e.g. only copy a word at a time when in
-  // little-endian mode).
-  for (size_t i = 0; i < len; ++i) {
+  // Copy message using aligned word sized loads and stores where possible to
+  // improve performance. Note: the parts of the message copied a byte at a time
+  // will not be byte swapped in big-endian mode.
```

Comment:
Oof! Would it make more sense to just disallow misaligned messages in big-endian mode?

Reply:
> Would it make more sense to just disallow misaligned messages in big-endian mode?

Maybe. The current approach does provide a little bit more flexibility. It means that you can still write arbitrary messages to the FIFO by calling `dif_kmac_absorb` multiple times (e.g. to add a byte-sized tag at the end of a message or something like that). Big-endian mode seems very specialized though so I don't really know if we care.

> Note that you can zero-pad the trailing bytes easily enough, so it's really just the leading ones that would be a problem.

Zero padding messages would change the length of the message and therefore the resulting digest so I'm not sure I follow.

Reply:
About the zero padding: Yes, you're completely right!

For the general question: It seems like "reverse the bytes for some of my buffer" is a seriously weird thing to want to do, and almost certainly implies a programming error. My preference would probably be to disallow messages in big-endian mode unless they are aligned and have length a multiple of 4. We can always change the API to allow it in future if someone thinks of a use case, but removing stuff from APIs is harder.

Reply:
> (Just a comment from the sidelines: we can also change the hardware if it makes things easier/more intuitive for software!)

Personally I'd just get rid of the option and let software byte swap if needed. I am not the end consumer though so there may be a use case that I am not aware of.

Reply:
@eunchan Do you remember where the endian swapping requirement came from in KMAC?

@mundaym can ask around in the SW meeting tonight if somebody on the SW side knows the if/why endian swapping might be required.

Reply:
In HMAC, if I recall correctly, we ended up using it to match the OpenSSL format.  I tried playing with SHA3, and it seems to match the output in LE config mode.

```cc
    {
        // $ echo -n "01234567" | openssl dgst -sha3-256
        // (stdin) = 3f1347c106caaf7be58f3fc4f145bc6c2588fc1a5e85d0a2d27df55673d09d5d
        .mode = kDifKmacModeSha3Len256,
        .message = "01234567",
        .message_len = 8,
        .digest = {
          0xc147133f,
          0x7bafca06,
          0xc43f8fe5,
          0x6cbc45f1,
          0x1afc8825,
          0xa2d0855e,
          0x56f57dd2,
          0x5d9dd073,
          },
        .digest_len = DIGEST_LEN_SHA3_256,
    },
```